### PR TITLE
feat: add Backstage Go coverage reporting (and gcov2lcov tool)

### PR DESCRIPTION
**Pull request by maxekman**
Add Backstage coverage reporting command and gcov2lcov tool needed for it.

https://github.com/jandelgado/gcov2lcov

### CODE REVIEW

File: tools/sgbackstagecoverage/tools.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+	"go.einride.tech/sage/sg"
+	"go.einride.tech/sage/tools/sggcov2lcov"
+)
+
+func PostCoverage(ctx context.Context, entity string) error {
+	if err := sggcov2lcov.Command(ctx).Run(); err != nil {
+		return fmt.Errorf("converting coverage format from gcov to lcov: %w", err)
+	}
+	url := fmt.Sprintf("https://api.roadie.so/api/code-coverage/report?entity=%s&coverageType=lcov", entity)
```
Comment (maxekman) [inline tools/sgbackstagecoverage/tools.go:17]: Should the Backstage URL be a parameter here?

Comment (maxekman) [reply tools/sgbackstagecoverage/tools.go]: Made it a parameter, together with the auth env.

File: tools/sggcov2lcov/tools.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+const (
+	name    = "gcov2lcov"
+	version = "1.0.6"
+)
+
+func Command(ctx context.Context, args ...string) *exec.Cmd {
+	sg.Deps(ctx, PrepareCommand)
+	coverFile := sg.FromBuildDir("go", "coverage", "go-test.txt")
+	lcovFile := sg.FromBuildDir("go", "coverage", "go-test.lcov")
```
Comment (alethenorio) [inline tools/sggcov2lcov/tools.go:24]: The convention in Sage is generally to keep the command as clean as possible running nothing but the tool and leaving any conventions with the tool to more specific commands. Can we leave these out of this command?

File: tools/sggcov2lcov/tools.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		append(
+			[]string{
+				"-infile=" + coverFile,
+				"-outfile=" + lcovFile,
+			},
+			args...,
+		)...,
+	)
+	goroot := sg.Output(sg.Command(ctx, "go", "env", "GOROOT"))
+	cmd.Env = append(cmd.Env, "GOROOT="+goroot)
```
Comment (alethenorio) [inline tools/sggcov2lcov/tools.go:37]: Why is this needed? Setting the GOROOT is something one should generally not need

File: tools/sgbackstagecoverage/tools.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+	"go.einride.tech/sage/sg"
+	"go.einride.tech/sage/tools/sggcov2lcov"
+)
+
+// PostGoCoverage posts Go code coverage to the Backstage API.
+// apiURL is the URL to the API, for example 'https://api.mybackstagedeployment.com'.
+// entity is the entity ref, for example 'component:default/my-service'.
+// authEnv is the env var with the Bearer, for example 'MY_BACKSTAGE_TOKEN'.
+func PostGoCoverage(ctx context.Context, apiURL, entity, authEnv string) error {
```
Comment (alethenorio) [inline tools/sgbackstagecoverage/tools.go:17]: What do you think about passing the token directly and letting the caller decide whether that lives in an environment variable or a file or something else?

File: tools/sggcov2lcov/tools.go (hunk 2 above)
Comment (maxekman) [reply tools/sggcov2lcov/tools.go]: Good point, I have moved these to be arguments instead.

File: tools/sggcov2lcov/tools.go (hunk 3 above)
Comment (maxekman) [reply tools/sggcov2lcov/tools.go:35]: It was needed when I ran the command. Not sure about the details, but it was mentioned in the tool docks so there is probably a valid reason: https://github.com/jandelgado/gcov2lcov#goroot

File: tools/sgbackstagecoverage/tools.go (hunk 4 above)
Comment (maxekman) [reply tools/sgbackstagecoverage/tools.go]: Moved it to an argument instead, and also removed the dependency on the `sggcov2lcov` command making this a pure LCOV post command.

File: tools/sggcov2lcov/tools.go (hunk 3 above)
Comment (alethenorio) [reply tools/sggcov2lcov/tools.go:35]: Did we try without it? The docs mention it might be needed and it would be great to skip it as it could lead to issues in some setups.

Comment (maxekman) [reply tools/sggcov2lcov/tools.go:35]: Yes, otherwise I would not have added it.

Comment (maxekman) [reply tools/sggcov2lcov/tools.go:35]: Do you have an idea how we can make it optional instead if you prefer?

File: tools/sggcov2lcov/tools.go (hunk 2 above)
Comment (alethenorio) [reply tools/sggcov2lcov/tools.go]: The command  target should be a clean target and run nothing else but the command and given args. No assumptions about any flags should be made in here.

Either we should use the infile and outfile arguments locally in each project or break it out to a separate target

File: tools/sgbackstagecoverage/tools.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	"net/http"
+	"os"
+)
+
+// PostCoverage posts code coverage in LCOV format to the Backstage API.
+// apiURL is the URL to the API, for example 'https://api.mybackstagedeployment.com'.
+// authToken is the Bearer token for the Backstage API.
+// coverFile is the coverage file in LCOV format (use sggcov2lcov to convert Go test output).
+// entity is the entity ref, for example 'component:default/my-service'.
+func PostCoverage(ctx context.Context, apiURL, authToken, coverFile, entity string) error {
```
Comment (alethenorio) [inline tools/sgbackstagecoverage/tools.go:15]: We should keep in mind Go readability conventions here. With this target name and package name we'll get `sgbackstagecoverage.PostCoverage`. Maybe rename the target to just `Post`?

Comment (maxekman) [reply tools/sgbackstagecoverage/tools.go:15]: Sure, good idea.

File: tools/sggcov2lcov/tools.go (hunk 3 above)
Comment (alethenorio) [reply tools/sggcov2lcov/tools.go:35]: You can let the user simply set the environment variable on the local project. Will the tool even work without it though? I am mostly trying to avoid future issues. Having to set GOROOT feels a little like working against the Go ecosystem but if the tool requires it to work I don't see any harm in keeping it for now. Maybe add a comment on it?

**Review by alethenorio — APPROVED** [review]
